### PR TITLE
story(notifications): support apache kafka

### LINKED ISSUE #24: story(notifications): support apache kafka
### Description

As an end user, I want to subscribe to notifications in Kafka, so that my infrastructure can be cloud provider agnostic.

### Acceptance Criteria

- [ ] Notifications can be published to Kafka

### Related Issues

_No response_

### PR DESCRIPTION
**Pull request by Zaba505**
Closes #24

### CODE REVIEW

File: notification/bus.go [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+	evryspb "github.com/z5labs/evrys/proto"
+)
+
+var (
+	ErrMissingLogger = errors.New("logger is required")
+)
+
+// Bus
+type Bus interface {
+	// Publish
```
Comment (erictg) [inline notification/bus.go:33]: more descriptive comment needed

File: notification/bus.go [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+	"strings"
+
+	evryspb "github.com/z5labs/evrys/proto"
+)
+
+var (
+	ErrMissingLogger = errors.New("logger is required")
+)
+
+// Bus
```
Comment (erictg) [inline notification/bus.go:31]: more descriptive comment needed

File: notification/bus.go [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+	Publish(context.Context, *evryspb.Notification) error
+}
+
+// Error represents a generic error a Bus implementation encountered
+type Error struct {
+	Bus   string
+	Cause error
+}
+
+func (e Error) Error() string {
```
Comment (erictg) [inline notification/bus.go:43]: need comment for exported function

File: notification/bus.go [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+	return fmt.Sprintf("%s: encountered unexpected error: %s", e.Bus, e.Cause)
+}
+
+func (e Error) Unwrap() error {
```
Comment (erictg) [inline notification/bus.go:47]: need comment for exported function

File: notification/bus.go [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+	return e.Cause
+}
+
+// MarshalError
+type MarshalError struct {
+	Protocol string
+	Cause    error
+}
+
+func (e MarshalError) Error() string {
```
Comment (erictg) [inline notification/bus.go:57]: need comment for exported function

File: notification/bus.go [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+	return fmt.Sprintf("%s: failed to marshal notification: %s", e.Protocol, e.Cause)
+}
+
+func (e MarshalError) Unwrap() error {
```
Comment (erictg) [inline notification/bus.go:61]: need comment for exported function

File: notification/bus.go [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+	return e.Cause
+}
+
+// ConfigurationError
+type ConfigurationError struct {
+	Bus   string
+	Cause error
+}
+
+func (e ConfigurationError) Error() string {
```
Comment (erictg) [inline notification/bus.go:71]: need comment for exported function

File: notification/bus.go [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+	return fmt.Sprintf("%s: invalid configuration: %s", e.Bus, e.Cause)
+}
+
+func (e ConfigurationError) Unwrap() error {
```
Comment (erictg) [inline notification/bus.go:75]: need comment for exported function

File: notification/bus.go [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+	return e.Cause
+}
+
+// ValidationError
+type ValidationError struct {
+	Cause error
+}
+
+func (e ValidationError) Error() string {
```
Comment (erictg) [inline notification/bus.go:84]: need comment for exported function

File: notification/bus.go [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+	return fmt.Sprintf("notification: must provide a valid notification: %s", e.Cause)
+}
+
+func (e ValidationError) Unwrap() error {
```
Comment (erictg) [inline notification/bus.go:88]: need comment for exported function

File: notification/bus.go [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+	return e.Cause
+}
+
+var (
+	ErrInvalidEventType   = errors.New("event type must be non-empty and non-blank")
+	ErrInvalidEventId     = errors.New("event id must be non-empty and non-blank")
+	ErrInvalidEventSource = errors.New("event source must be non-empty and non-blank")
```
Comment (erictg) [inline notification/bus.go:95]: need comments for exported types

File: notification/kafka.go [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+	evryspb "github.com/z5labs/evrys/proto"
+
+	"github.com/segmentio/kafka-go"
+	"go.uber.org/zap"
+	"google.golang.org/protobuf/proto"
+)
+
+var kafkaBus = "kafka"
+
+// KafkaBus
```
Comment (erictg) [inline notification/kafka.go:30]: need more descriptive comment

File: notification/kafka.go [hunk 13]
```diff
@@ -0,0 +1,151 @@
...
+type KafkaBus struct {
+	kafka *kafka.Writer
+
+	log *zap.Logger
+}
+
+// KafkaConfig
```
Comment (erictg) [inline notification/kafka.go:37]: need more descriptive comment

File: notification/kafka.go [hunk 14]
```diff
@@ -0,0 +1,151 @@
...
+type KafkaConfig struct {
+	Addresses              []string
+	AllowAutoTopicCreation bool
+	Logger                 *zap.Logger
+}
+
+var (
+	ErrMissingKafkaAddrs = errors.New("kafka addresses are required")
```
Comment (erictg) [inline notification/kafka.go:45]: need comment for exported value

File: notification/kafka.go [hunk 15]
```diff
@@ -0,0 +1,151 @@
...
+)
+
+// NewKafkaBus
+func NewKafkaBus(cfg KafkaConfig) *KafkaBus {
+	err := validateKafkaConfig(cfg)
+	if err != nil {
+		panic(ConfigurationError{
+			Bus:   kafkaBus,
+			Cause: err,
+		})
```
Comment (erictg) [inline notification/kafka.go:55]: panic bad, return an error instead of panicking

File: notification/kafka.go [hunk 16]
```diff
@@ -0,0 +1,151 @@
...
+}
+
+func validateKafkaAddrs(cfg KafkaConfig) error {
+	if cfg.Addresses == nil || len(cfg.Addresses) == 0 {
+		return ErrMissingKafkaAddrs
+	}
+	return nil
+}
+
+// Publish
```
Comment (erictg) [inline notification/kafka.go:95]: need more descriptive comment

File: notification/kafka.go [hunk 17]
```diff
@@ -0,0 +1,151 @@
...
+	kafka *kafka.Writer
+
+	log *zap.Logger
+}
+
+// KafkaConfig
+type KafkaConfig struct {
+	Addresses              []string
+	AllowAutoTopicCreation bool
+	Logger                 *zap.Logger
```
Comment (erictg) [inline notification/kafka.go:41]: why provide the logger via the config. Wouldn't it be better to use the global zap for consistency and just add required fields for kafka?

File: notification/kafka.go [hunk 18]
```diff
@@ -0,0 +1,151 @@
...
+type KafkaBus struct {
+	kafka *kafka.Writer
+
+	log *zap.Logger
+}
+
+// KafkaConfig
+type KafkaConfig struct {
+	Addresses              []string
+	AllowAutoTopicCreation bool
```
Comment (erictg) [inline notification/kafka.go:40]: add validator tags and use the validator for this

**Review by erictg — CHANGES_REQUESTED** [review]
move errors into a package level error file

File: notification/kafka.go (hunk 17 above)
Comment (Zaba505) [reply notification/kafka.go:41]: Global state should always be avoided. Only the `cmd` package is using the global logger and that's simply cuz I haven't updated it yet to not use the global one.

**Comment by Zaba505** [comment]
No longer going this route

**Pull request closed without merging**